### PR TITLE
Palette-based color picker

### PR DESCRIPTION
**Pull request by jamieparkinson**
This is an (arguably) improved UI for colour filtering based on the design [in Zeplin](https://zpl.io/V48dnYJ).

The library we're using doesn't export the plumbing that we want so I replicated it (ish) which makes the UI much better suited to our needs (and doesn't cause CSS import problems).

It looks like this:
image

### CODE REVIEW

File: common/views/components/PaletteColorPicker/HueSlider.tsx [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  background-color: white;
+  box-shadow: 0 0 1px rgba(black, 0.5);
+  border-radius: 2px;
+`;
+
+type InteractionEvent = MouseEvent | TouchEvent;
+type ReactInteractionEvent = React.MouseEvent | React.TouchEvent;
+
+const isTouch = (e: InteractionEvent): e is TouchEvent => 'touches' in e;
+const clamp = (x: number, min = 0, max = 1) =>
```
Comment (jamesgorrie) [inline common/views/components/PaletteColorPicker/HueSlider.tsx:56]: nit: We have `lodash.clamp`.

Comment (jamieparkinson) [reply common/views/components/PaletteColorPicker/HueSlider.tsx:56]: Do we? I didn't want to add a dependency for such a simple function

File: common/views/components/PaletteColorPicker/HueSlider.tsx [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  x > max ? max : x < min ? min : x;
+const useIsomorphicLayoutEvent =
```
[the remark this answers is not in the document]
Comment (jamieparkinson) [reply common/views/components/PaletteColorPicker/HueSlider.tsx:58]: Got it from here https://github.com/omgovich/react-colorful/blob/c4785feff95035f0f73244430dd2f22d351d7713/src/hooks/useIsomorphicLayoutEffect.ts

[the remark this answers is not in the document]
Comment (jamieparkinson) [reply common/views/components/PaletteColorPicker/HueSlider.tsx:58]: https://github.com/facebook/react/issues/14927

File: common/views/components/PaletteColorPicker/HueSlider.tsx (hunk 1 above)
Comment (jamesgorrie) [reply common/views/components/PaletteColorPicker/HueSlider.tsx:56]: Yea, we could put it in `utils/`

File: common/views/components/PaletteColorPicker/HueSlider.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (jamesgorrie) [reply common/views/components/PaletteColorPicker/HueSlider.tsx:58]: Interesting, I wonder if we could just use `useLayoutEffect` though. Happy for either.

File: common/views/components/PaletteColorPicker/PaletteColorPicker.tsx [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  margin-top: 15px;
+`;
+
+const ColorLabel = styled.label<{ active: boolean }>`
+  font-style: italic;
+  color: ${({ active }) => (active ? '#121212' : '#565656')};
+  font-size: 14px;
+`;
+
+const ClearButton = styled.button`
```
Comment (jamesgorrie) [inline common/views/components/PaletteColorPicker/PaletteColorPicker.tsx:55]: Does this make them submit buttons?

**Review by jamesgorrie — APPROVED** [review]
Loving this - I also like that this could easily be implemented sans-JS.

File: common/views/components/PaletteColorPicker/PaletteColorPicker.tsx (hunk 3 above)
Comment (jamieparkinson) [reply common/views/components/PaletteColorPicker/PaletteColorPicker.tsx:55]: good point, will check

**Pull request merged**